### ISSUE TITLE
"Unused bucklescript attribute" error not parsed

### ISSUE THREAD

**Issue by jchavarri** [freebroccolo/ocaml-language-server#83]
Another case where the `bsb` regexps don't work:

```
type fruit = [
  | `Apple
  | [@bs.as "miniCoconut"] `Kiwi
  | `Watermelon
];
```

`bsb` output in console:
```
ninja: Entering directory `lib/bs'
[1/2] Building src/Demo.mlast
FAILED: src/Demo.mlast 
/Users/javi/.nvm/versions/node/v6.11.2/lib/node_modules/bs-platform/lib/bsc.exe -pp "/Users/javi/.nvm/versions/node/v6.11.2/lib/node_modules/bs-platform/lib/refmt3.exe --print binary"    -w -30-40+6+7+27+32..39+44+45+101 -warn-error +101 -bs-suffix -nostdlib -I '/Users/javi/Development/sandbox/my-first-app/node_modules/bs-platform/lib/ocaml' -no-alias-deps -color always -c -o src/Demo.mlast -bs-syntax-only -bs-binary-ast -impl /Users/javi/Development/sandbox/my-first-app/src/Demo.re
File "/Users/javi/Development/sandbox/my-first-app/src/Demo.re", line 7, characters 6-11:
Warning 101: Unused bucklescript attribute: bs.as
File "/Users/javi/Development/sandbox/my-first-app/src/Demo.re", line 1:
Error: Some fatal warnings were triggered (1 occurrences)
ninja: error: rebuilding 'build.ninja': subcommand failed
```

@freebroccolo @chenglou @bobzhang not sure about the best way to fix these. Should we:

- keep trying to add more regexps to the existing `bsb` parser?
- use chevrotain like @freebroccolo suggested in https://github.com/freebroccolo/ocaml-language-server/issues/51#issuecomment-352298650?
- uniforming the output from BuckleScript, like the approach attempted in https://github.com/BuckleScript/bucklescript/pull/2421?
- Or is it smarter to just focus on building a lang server mode for `bsb`? Happy to contribute if there are some initial pointers on how to start working on this :)

**Comment by freebroccolo** [comment]
> keep trying to add more regexps to the existing bsb parser?

In the short term the easiest solution is going to be to adjust the regexps in the `bsb` parser. 

> use chevrotain like @freebroccolo suggested in #51 (comment)?

In the long term I think rewriting the parsers with chevrotain or something similar is going to be necessary just to keep the server maintainable. However, the native rewrite of the server itself might happen before then, in which case we'd want to use Menhir.

> uniforming the output from BuckleScript, like the approach attempted in BuckleScript/bucklescript#2421?

I don't have much to say about upstream changes but that might be helpful. In general, if it were possible to get the bucklescript tooling to output JSON (maybe with a special flag) that would make things a lot easier for us.

> Or is it smarter to just focus on building a lang server mode for bsb? Happy to contribute if there are some initial pointers on how to start working on this :)

I'm not sure quite what you have in mind for this. I don't think it's really necessary to try to have like a completely separate analyzer class just for `bsb` but there certainly is room for cleaning up the current behavior of the server. The main thing I'd want to avoid would be duplication of features or things like diagnostic handling logic.

**Comment by freebroccolo** [comment]
@jchavarri do you think it is okay to just go with some additional regexps here for the time being? We could revisit this more thoroughly after some of the bigger upcoming changes land in the master branch.

**Comment by jchavarri** [comment]
@freebroccolo 👍  totally. I think I'll do this one and #51 in one PR. I was trying to fix #51 upstream in https://github.com/BuckleScript/bucklescript/pull/2421 but i couldn't find a way.

Also, I've been thinking about it and i'd like to take this opportunity and add some tests for this parser as well, so we can add cases like this. What I thought was going to be a temporary hack to parse `bsb` output seems to be much more permanent, which is good! :) i just feel the current setup for the parser is quite brittle.

Jest would be my preferred choice, but let me know if you have other ideas.

**Comment by jchavarri** [comment]
> > Or is it smarter to just focus on building a lang server mode for bsb? Happy to contribute if there are some initial pointers on how to start working on this :)

> I'm not sure quite what you have in mind for this.

I meant having `bsb` run in a process like Merlin does. I remember @chenglou mentioning at some point that this "process-based `bsb`" could be a thing, although the current approach of just calling it every time works fine too, and its simplicity is also something to take into account.
